packages/jss/src/DomRenderer.js

```diff
@@ -38,19 +38,23 @@ const setProperty = (cssRule, prop, value) => {
     let cssValue = value
 
     if (Array.isArray(value)) {
-      cssValue = toCssValue(value, true)
-
-      if (value[value.length - 1] === '!important') {
-        cssRule.style.setProperty(prop, cssValue, 'important')
-        return true
-      }
+      cssValue = toCssValue(value)
     }
 
+    const indexOfImportantFlag = cssValue ? cssValue.indexOf('!important') : -1
+
+    const cssValueWithoutImportantFlag =
+      indexOfImportantFlag > -1 ? cssValue.substr(0, indexOfImportantFlag - 1) : cssValue
```

Comment:
2 things for potential improvement:

1. need to compare if there is a serious advantage in performance: .indexOf + .substr vs .replace, because you could be replacing instead of searching and then .substr, but I am not sure there is a perf difference, we need an isolated bench test to see if there is a big difference, I am usually using https://esbench.com/
2. In any case we only need this on line 55, inside the else, so we are not running it in modern browsers which support attributeStyleMap, so we need to move it down

Reply:
The reason its important to move it down into else is that this is being called very frequently and if we can avoid doing that string manipulation on every value - we should

Reply:
What about this code?

```js
let includedImportantFlag = false
const cssValueWithoutImportantFlag = cssValue
  ? cssValue.replace('!important', () => {
      includedImportantFlag = true
      return ''
    })
  : cssValue

if (cssRule.attributeStyleMap) {
  cssRule.attributeStyleMap.set(prop, cssValue)
} else {
  cssRule.style.setProperty(
    prop,
    cssValueWithoutImportantFlag,
    includedImportantFlag ? 'important' : ''
  )
}
```

Reply:
or this:

```js
let priority = ''
const cssValueWithoutImportantFlag = cssValue
  ? cssValue.replace('!important', () => {
      priority = 'important'
      return ''
    })
  : cssValue

if (cssRule.attributeStyleMap) {
  cssRule.attributeStyleMap.set(prop, cssValue)
} else {
  cssRule.style.setProperty(prop, cssValueWithoutImportantFlag, priority)
}
``

Reply:
lets compare the perf on esbench between this and .indexOf version, we might be unaware of high difference and this is a hot path

Reply:
btw. on esbench I would only compare the string replacement in isolation, nothing else

Reply:
I tested by this sample:

# indexOf + substr
```js
function run(value){
  const index = value.indexOf("!important")

  const result = value.substr(0, index);
}

run("background-color:red !important;");
run("color:rbg(0, 0, 0) !important;");
run("background-color:rbg(0, 0, 0);");
```

# replace
```js
function run(value){
  const result = value.replace("!important", "");
}

run("background-color:red !important;");
run("color:rbg(0, 0, 0) !important;");
run("background-color:rbg(0, 0, 0);");
```

And the result is:
<img width="775" alt="Screenshot 2022-04-24 at 23 27 54" src="https://user-images.githubusercontent.com/1549069/164997431-f0858b92-0f22-4f2b-8001-c906d0a540fa.png">